### PR TITLE
feat: recovery dispatch notifications

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

This adds recovery notifications and in doing so [hides the proposal widget when a proposal is in progress](https://www.notion.so/safe-global/Show-Account-recovery-in-progress-immediately-57206a130c004fa58f5425c91e6605a0).

## How this PR fixes it

This builds upon the recovery event bus, adding notifications for Recoverer recovery transactions as we do for "standard" transactions and messages:

- Recovery proposal
- Recovery execution
- Skipping of expired recovery attempts

Each of the above have notifications as such throughout their lifecycle:

1. "Confirm the execution in your wallet..."
2. "Validating..."
3. "Successfully validated. Querying..."

If the transaction reverts or fails, the following is displayed accordingly:

1. "Reverted. Please check your gas settings."
2. "Failed."

With the implementation of the above, it is now possible to know when a proposal is "in progress", e.g. being submitted to or validated on-chain. As such, the dashboard widget for Recoverers that allows for starting recovery is hidden when a proposal is "in progress" (see recording below).

## How to test it

1. Propose a recovery and observe the new notifications, but also the "Start recovery" header hidden during the proposal lifecycle.
2. Execute a recovery and observe the new notifications.
3. Skip expired recovery attempts and observe the new notifications.

## Screenshots

image

recovery notifications

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/dashboard/RecoveryHeader/index.test.tsx [hunk 1]
```diff
@@ -46,22 +41,80 @@ describe('RecoveryHeader', () => {
...
+        recoveryTxHash: faker.string.hexadecimal(),
+        eventType: faker.helpers.enumValue(RecoveryEventType),
+      })
+
+      await waitFor(() => {
+        expect(result.current).toBe(true)
+      })
+    })
+  })
+  ;[RecoveryEvent.REVERTED, RecoveryEvent.PROCESSED, RecoveryEvent.FAILED].forEach((event) => {
```
[the remark this answers is not in the document]
Comment (schmanu) [reply src/components/dashboard/RecoveryHeader/index.test.tsx:91]: I figured it out. It is because the recovery did not hit the queue yet.
Could you add that to the description?

File: src/components/dashboard/RecoveryHeader/index.test.tsx [hunk 2]
```diff
@@ -46,22 +41,80 @@ describe('RecoveryHeader', () => {
...
+
+      const { result } = renderHook(() => _useIsProposalInProgress())
+
+      expect(result.current).toBe(false)
+
+      recoveryDispatch(event, {
+        moduleAddress: faker.finance.ethereumAddress(),
+        txHash: faker.string.hexadecimal(),
+        recoveryTxHash: faker.string.hexadecimal(),
+        eventType: faker.helpers.enumValue(RecoveryEventType),
```
Comment (schmanu) [inline src/components/dashboard/RecoveryHeader/index.test.tsx:83]: Shouldn't the eventType always be RecoveryEventType.Proposal?

File: src/components/dashboard/RecoveryHeader/index.test.tsx [hunk 3]
```diff
@@ -46,22 +41,80 @@ describe('RecoveryHeader', () => {
...
+      })
+    })
+  })
+  ;[RecoveryEvent.REVERTED, RecoveryEvent.PROCESSED, RecoveryEvent.FAILED].forEach((event) => {
+    it('should return false if there is not a proposal in progress', async () => {
+      const payload = {
+        moduleAddress: faker.finance.ethereumAddress(),
+        txHash: faker.string.hexadecimal(),
+        recoveryTxHash: faker.string.hexadecimal(),
+        eventType: faker.helpers.enumValue(RecoveryEventType),
```
Comment (schmanu) [inline src/components/dashboard/RecoveryHeader/index.test.tsx:97]: Same here. Shouldn't it always be RecoveryEventType.Proposal?

File: src/components/dashboard/RecoveryHeader/index.tsx [hunk 4]
```diff
@@ -46,11 +48,33 @@ export function _RecoveryHeader({
...
 }
 
+export function _useIsProposalInProgress(): boolean {
+  const [isProposalSubmitting, setIsProposalSubmitting] = useState(false)
+  const queue = useRecoveryQueue()
+
+  useEffect(() => {
+    const unsubFns = Object.values(RecoveryEvent).map((event) =>
+      recoverySubscribe(event, (detail) => {
+        const isValidating = event === RecoveryEvent.EXECUTING || event === RecoveryEvent.PROCESSING
```
Comment (schmanu) [inline src/components/dashboard/RecoveryHeader/index.tsx:58]: Don't we have to ignore all events with `detail.eventType !== RecoveryEventType.PROPOSAL`?

File: src/hooks/useRecoveryTxNotification.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    if (!chain?.blockExplorerUriTemplate) {
+      return
+    }
+
+    const unsubFns = Object.entries(RecoveryTxNotifications).map(([event, notification]) =>
+      recoverySubscribe(event as RecoveryEvent, async (detail) => {
+        const isSuccess = event === RecoveryEvent.PROCESSED
+        const isError = 'error' in detail
+
+        const txHash = 'txHash' in detail ? detail.txHash : undefined
```
Comment (schmanu) [inline src/hooks/useRecoveryTxNotification.ts:48]: Not sure where to put this comment.
But there is a edge case of having more than one expired txs.
Canceling one will actually cancel all of them. 
Not highest prio as it's an edge case but I think currently it would only mark the first expired recovery proposal as "processing" or being cancelled.

File: src/components/dashboard/RecoveryHeader/index.test.tsx (hunk 2 above)
Comment (iamacook) [reply src/components/dashboard/RecoveryHeader/index.test.tsx]: Adjusted in 03398c6976df9d148a727b5e15e8ba09feb7309f.

File: src/components/dashboard/RecoveryHeader/index.test.tsx (hunk 3 above)
Comment (iamacook) [reply src/components/dashboard/RecoveryHeader/index.test.tsx]: Adjusted in 03398c6976df9d148a727b5e15e8ba09feb7309f.

File: src/components/dashboard/RecoveryHeader/index.tsx (hunk 4 above)
Comment (iamacook) [reply src/components/dashboard/RecoveryHeader/index.tsx:59]: Adjusted in 03398c6976df9d148a727b5e15e8ba09feb7309f.

File: src/hooks/useRecoveryTxNotification.ts (hunk 5 above)
Comment (iamacook) [reply src/hooks/useRecoveryTxNotification.ts:49]: It's not possible to propose more than one recovery transaction anymore so I would not concern ourselves about this. Users need to cancel any expired recovery attempts before they can make a new one.
